### PR TITLE
Circulars: Open Automatically Formatted Hyperlinks in New Tab

### LINKED ISSUE #2389: Circulars: Open external links in separate tab
# Description
GCN parses Circulars to find commonly referenced sources (like TNS IDs, arXiv codes, DOIs, and other Circulars) and URLs to automatically convert them into hyperlinks and provide rich tooltip content without navigating from the page. 

In order to preserve continuity when browsing the circular archive, it would be beneficial to make it so external links open in a new tab but have circulars open in the same tab due to being internal links.

# Acceptance criteria

- [x] This change should not impact the parsing and replacement of referenced sources and URLs
- [x] Any links to other circulars should open in the same tab 
- [x] All external links should open in a new tab

### PR DESCRIPTION
**Pull request by tylerbarna**
# Description
This makes any plaintext hyperlinks open in a new tab rather than opening in the same tab.

However, this currently does not do the same for tns, doi, arxiv, or other circular links formatted by remark-rehype-astro. It might make sense to make this same change for tns, doi, and arxiv links but retain the current behavior for automatic circular links since that is intrasite navigation.

# Related Issue(s)
Resolves #2389  

# Testing
1. Navigate to any Circular that contains a URL
2. Click on the URL. This should now open it in a new tab rather than the same tab

### CODE REVIEW

**Review by dakota002 — APPROVED** [review]

File: app/routes/circulars.$circularId.($version)/rehypeAutolinkLiteral.ts [hunk 1]
```diff
@@ -20,7 +20,11 @@ const regexp = new RegExp(
 function autolinkLiteral(tree: Parameters<typeof findAndReplace>[0]) {
   findAndReplace(
     tree,
-    [regexp, (href: string) => h('a', { rel: 'external', href }, href)],
+    [
+      regexp,
+      (href: string) =>
+        h('a', { rel: 'external', href, target: '_blank' }, href),
```
Comment (lpsinger) [inline app/routes/circulars.$circularId.($version)/rehypeAutolinkLiteral.ts:26]:
```suggestion
        h('a', { rel: 'external', href, target: '_blank', rel='noreferrer' }, href),
```

**Review by lpsinger — CHANGES_REQUESTED** [review]
I'd like to keep all of the hyperlinks from GCN Circulars consistent: links to other Circulars should open in the same page; links to external pages can open in new windows or tabs. Would it be possible to make that change in one single PR so that it is never inconsistent?

File: app/routes/circulars.$circularId.($version)/rehypeAutolinkLiteral.ts (hunk 1 above)
Comment (tylerbarna) [reply app/routes/circulars.$circularId.($version)/rehypeAutolinkLiteral.ts:26]: tweaking suggestion in its own commit to prevent property from being repeated

**Comment by tylerbarna** [comment]
> I'd like to keep all of the hyperlinks from GCN Circulars consistent: links to other Circulars should open in the same page; links to external pages can open in new windows or tabs. Would it be possible to make that change in one single PR so that it is never inconsistent?

I agree on the point about consistency; I'm working on modifying the AstroDataLink component now

File: app/routes/circulars.$circularId.($version)/AstroDataContext.tsx [hunk 2]
```diff
@@ -31,16 +31,23 @@ export const AstroDataContext = createContext<AstroDataContextProps>({})
  */
 export const AstroDataLink = forwardRef(
   (
-    { children, className, rel: origRel, ...props }: Omit<LinkProps, 'target'>,
+    {
+      children,
+      className,
+      rel: origRel,
+      isExternal,
```
Comment (lpsinger) [inline app/routes/circulars.$circularId.($version)/AstroDataContext.tsx:38]: Could we please call this prop `external`?

File: app/routes/circulars.$circularId.($version)/AstroDataContext.tsx [hunk 3]
```diff
@@ -31,16 +31,23 @@ export const AstroDataContext = createContext<AstroDataContextProps>({})
...
+      ...props
+    }: Omit<LinkProps, 'target'> & { isExternal?: boolean },
     ref: Ref<HTMLAnchorElement>
   ) => {
     const context = useContext(AstroDataContext)
+    const target = isExternal ? '_blank' : context.target
     const rel = [origRel, context.rel].filter(Boolean).join(' ') || undefined
```
Comment (lpsinger) [inline app/routes/circulars.$circularId.($version)/AstroDataContext.tsx:45]: Note that we also need to add `'external noopener'` if the link is external.

File: app/routes/circulars.$circularId.($version)/AstroDataContext.tsx [hunk 4]
```diff
@@ -62,11 +69,13 @@ export const AstroDataLink = forwardRef(
 export function AstroDataLinkWithTooltip<T>({
   fetch,
   label,
+  isExternal,
```
Comment (lpsinger) [inline app/routes/circulars.$circularId.($version)/AstroDataContext.tsx:72]: You can safely remove this; it will be passed down along with `...props`.

```suggestion
```

File: app/routes/circulars.$circularId.($version)/rehypeAutolinkLiteral.ts [hunk 5]
```diff
@@ -20,7 +20,11 @@ const regexp = new RegExp(
 function autolinkLiteral(tree: Parameters<typeof findAndReplace>[0]) {
   findAndReplace(
     tree,
-    [regexp, (href: string) => h('a', { rel: 'external', href }, href)],
+    [
+      regexp,
+      (href: string) =>
+        h('a', { rel: 'noreferrer', href, target: '_blank' }, href),
```
Comment (lpsinger) [inline app/routes/circulars.$circularId.($version)/rehypeAutolinkLiteral.ts:26]: Note that there is a `LinkWrapper` component in `app/routes/circulars.$circularId.($version)/Body.tsx` with which we replace all Markdown links. Please just set your new `isExternal` or `external` prop in that component, then you can undo this change:

```suggestion
        h('a', { target: '_blank' }, href),
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/circulars.$circularId.($version)/rehypeAutolinkLiteral.ts (hunk 5 above)
Comment (tylerbarna) [reply app/routes/circulars.$circularId.($version)/rehypeAutolinkLiteral.ts:26]: ah, I missed that, thanks!

**Comment by tylerbarna** [comment]
@lpsinger does #2405 and #2406 make this PR redundant?

**Comment by lpsinger** [comment]
> @lpsinger does #2405 and #2406 make this PR redundant?

No.

**Comment by tylerbarna** [comment]
> > @lpsinger does #2405 and #2406 make this PR redundant?
> 
> No.

ah okay, just wanted to make sure

**Comment by lpsinger** [comment]
Please address the remaining comments.

**Review by lpsinger — APPROVED** [review]

**Review by lpsinger — CHANGES_REQUESTED** [review]
Please fix TypeScript errors.

**Review by lpsinger — APPROVED** [review]
